tests/30rooms/06membership.pl

```diff
@@ -205,3 +206,157 @@
          );
       });
    };
+
+test "Can invite existing 3pid",
+   requires => [qw( do_request_json more_users test_http_server_uri_base await_http_request )],
+
+   do => sub {
+      my ( $do_request_json, $more_users, $test_http_server_uri_base, $await_http_request ) = @_;
+
+      my $invitee_email = "marmosets\@monkeyworld.org";
+      my $invitee_mxid = $more_users->[0]->user_id;
+      my $room_id;
+
+      Future->needs_all(
+         stub_is_lookup($invitee_email, $invitee_mxid, $await_http_request),
+
+         make_inviteonly_room($do_request_json)
+         ->on_done( sub {
+            ( $room_id ) = @_;
+         })->then( sub {
+            $do_request_json->(
+               method => "POST",
+               uri    => "/api/v1/rooms/$room_id/invite",
+
+               content => {
+                  id_server => (substr $test_http_server_uri_base, length("https://")),
+                  medium => "email",
+                  address => $invitee_email,
+                  display_name => "Cute things",
+               },
+            ),
+         })->then( sub {
+            $do_request_json->(
+               method => "GET",
+               uri    => "/api/v1/rooms/$room_id/state/m.room.member/$invitee_mxid",
+            )->on_done( sub {
+               my ( $body ) = @_;
+               log_if_fail $body;
+               $body->{membership} eq "invite" or
+                  die "Expected invited user membership to be 'invite'";
+            }),
+         }),
+      );
+   };
+
+sub stub_is_lookup {
+   my ( $email, $mxid, $await_http_request ) = @_;
+   $await_http_request->("/_matrix/identity/api/v1/lookup", sub {
+      my ( $req ) = @_;
+      return unless $req->query_param("medium") eq "email";
+      return unless $req->query_param("address") eq $email;
+      return 1;
+   })->then( sub {
+      my ( $request ) = @_;
+      my $content = ( defined($mxid) ? qq({"medium": "email", "address": "$email", "mxid": "$mxid"}) : qq({}) );
+      $request->respond(make_200($content));
+      Future->done( 1 );
+   })
+};
+
+sub make_200 {
+   my ( $content ) = @_;
+   HTTP::Response->new( 200, "OK", ["Content-Length", length($content)], $content );
+};
+
+sub make_inviteonly_room {
```

Comment:
Rather than hide this inbetween the tests, it might be less surprising to the reader encountering it on line 221 if it were higher up; before the test that uses it. Or put it right up at the top of the file.